### PR TITLE
[SPARK-25258][SPARK-23131][SPARK-25176][BUILD] Upgrade Kryo to 4.0.2

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

Upgrade chill to 0.9.3, Kryo to 4.0.2, to get bug fixes and improvements. 

The resolved tickets includes:
- SPARK-25258 Upgrade kryo package to version 4.0.2
- SPARK-23131 Kryo raises StackOverflow during serializing GLR model
- SPARK-25176 Kryo fails to serialize a parametrised type hierarchy

More details:
https://github.com/twitter/chill/releases/tag/v0.9.3
https://github.com/twitter/chill/commit/cc3910d501a844f3c882249fef8fc2560b95b6dd

## How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by wangyum** [comment]
cc @srowen

**Comment by srowen** [comment]
That looks like a major version bump -- the usual question here -- what are the key changes we need, what are possible incompatible changes?

**Comment by wangyum** [comment]
Thanks @srowen [SPARK-25176](https://issues.apache.org/jira/browse/SPARK-25176) has a detail description:

> I'm using the latest spark version spark-core_2.11:2.3.1 which 
transitively depends on com.esotericsoftware:kryo-shaded:3.0.3 via the
com.twitter:chill_2.11:0.8.0 dependency. This exact version of kryo serializer contains an issue [1,2] which results in throwing ClassCastExceptions when serialising parameterised type hierarchy.
This issue has been fixed in kryo version 4.0.0 [3]. It would be great to have this update in Spark as well. Could you please upgrade the version of com.twitter:chill_2.11 dependency from 0.8.0 up to 0.9.2?
You can find a simple test to reproduce the issue [4].
[1] https://github.com/EsotericSoftware/kryo/issues/384
[2] https://github.com/EsotericSoftware/kryo/issues/377
[3] https://github.com/EsotericSoftware/kryo/releases/tag/kryo-parent-4.0.0
[4] https://github.com/mpryahin/kryo-parametrized-type-inheritance

**Comment by srowen** [comment]
https://github.com/EsotericSoftware/kryo/releases/tag/kryo-parent-4.0.0

This has a number of breaking changes. However, Spark actually doesn't use kryo itself, and excludes it. It uses kryo-shaded, via chill. chill otherwise has almost no changes from 0.8 to 0.9 except to accommodate Kryo 4. Therefore the update shouldn't actually force downstream apps to use a different kryo, which is most of the potential problem here.

The changes in serialized format should not be visible outside Spark, if I am not mistaken. It controls serialization within Spark components and this isn't compatible across versions anyway.

I'm inclined to merge this for 2.4 as it resolves a number of issues users are hitting.

**Comment by erikerlandson** [comment]
If it doesn't expose any breaking changes to users, that seems reasonable. I assume there would be no integration problems with scala 2.12?

**Comment by srowen** [comment]
Kryo 3.0.3 and Chill 0.8.2 already work with 2.12, so no issue particularly there. I don't see that this change would be visible to users, and tests suggest it doesn't affect Spark. I'm cautious that I'm maybe overlooking something though. And weighing the risk vs the upside of fixing some user issues and potential performance bumps.

File: pom.xml [hunk 1]
```diff
@@ -1770,6 +1770,10 @@
             <groupId>org.apache.hive</groupId>
             <artifactId>hive-storage-api</artifactId>
           </exclusion>
+          <exclusion>
+            <groupId> com.esotericsoftware</groupId>
+            <artifactId>kryo-shaded</artifactId>
```
Comment (cloud-fan) [inline pom.xml]: why this change?

Comment (srowen) [reply pom.xml]: CC @wangyum yes good point. ORC also uses kryo-shaded, and uses 3.0.3. In theory that could cause a break, but the tests pass. That's a positive sign but not bulletproof. @dongjoon-hyun do you have any insight into how ORC uses kryo? Is it a code path that wouldn't matter to Spark?

Comment (dongjoon-hyun) [reply pom.xml]: Thank you for pinging me, @srowen . ORC uses Kryo only for writing/reading one ORC configuration, `orc.kryo.sarg`. The followings are the Spark's code indirect code path. I guess Kryo provides backward compatibility at lest, but I'll take a look at this PR today.
- For new ORCFileFormat, [OrcInputFormat.setSearchArgument](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala#L159)
- For old ORCFileFormat, [SearchArgument.toKryo](https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala#L129)

**Comment by erikerlandson** [comment]
For maximum breaking-change forgiveness, the "safe" option would be to punt it to 3.0; which seems likely to be the next release after this.

**Comment by dongjoon-hyun** [comment]
@wangyum . I know this is designed for a build-only PR. However, since this is claimed for bug fixes(SPARK-20389, SPARK-23131, and SPARK-25176), could you add test cases for them in order to verify those issues are really resolved?

File: pom.xml (hunk 1 above)
Comment (dongjoon-hyun) [reply pom.xml]: @srowen  In short, the current Spark always uses the same Kryo version for read/write `SearchArgument` and it's used only on runtime.

1. Old OrcFileFormat always uses `org.spark-project.hive:hive-exec:1.2.1.spark2` which uses the shaded one in `hive-exec`.
    - `com.esotericsoftware.kryo:kryo:2.21`. 

2. New OrcFileFormat uses `org.apache.orc` which uses the one provided by Spark.
    - `com.esotericsoftware:kryo-shaded:3.0.3`  (All Spark/Orc/Hive uses this version for now)

3. New OrcFileFormat (in this PR) uses `org.apache.orc` which uses the one provided by Spark.
    - `com.esotericsoftware:kryo-shaded:4.0.2`

So, (1) is unchanged by this PR. (2) and (3) also doesn't use a mixed version of Kryo. So, it should be fine because Apache Spark doesn't allow a mixed Spark version(master and executor). BTW, during investigation, there was some performance issue in `createFilter`. I'll file a new JIRA for that.

Comment (srowen) [reply pom.xml]: I guess my question is whether Orc complied against kryo-shaded 3.x necessarily works with kryo-shaded 4.x, because there were API and behavior changes. Our current tests don't seem to surface any such problems, but who knows if there's something they don't test.

I agree that our distribution won't include two versions, but the issue I'm wondering about is whether Orc will like the one later version that it gets bundled with.

Comment (dongjoon-hyun) [reply pom.xml]: Yes. I checked that, @srowen . `org.apache.orc` only uses Kryo constructor, `writeObject`, and `readObject` from `kryo-shaded` library. There is no change for them.

**WRITE**
```
(new Kryo()).writeObject(out, sarg);
```
**READ**
```
... = (new Kryo()).readObject(new Input(sargBytes), SearchArgumentImpl.class);
```

Comment (srowen) [reply pom.xml]: OK, if that's the extent of the usage, then I believe that ORC is OK with Kryo 4. That much is not a problem.

Comment (dongjoon-hyun) [reply pom.xml]: I was also worried that part, but it's used only in a run-time `SearchArgument` serialization. There was no usage with ORC `file`s.

**Comment by dongjoon-hyun** [comment]
Ping, @wangyum . Could you add the test cases for SPARK-20389, SPARK-23131, and SPARK-25176?

**Comment by wangyum** [comment]
@dongjoon-hyun I'm trying to add test cases.

**Comment by srowen** [comment]
Tests are great of course; those three might be hard to test. The first and last one don't have info on the reproduction. The second one looks possibly reproducible though as it shows the custom serialization code used to write a GeneralizedLinearRegression model.

**Comment by dongjoon-hyun** [comment]
I pinged the first one, SPARK-20389, to find any chance to test on the original cluster. Although it's too old issue, but there is no harm for us to ping once more.

For the third one, SPARK-25176, the reporter provided the sample code. Although the code is open for us, but it's not Apache license. We can make our own test case for type-inheritance.
> You can find a simple test to reproduce the issue [4].
- [4] https://github.com/mpryahin/kryo-parametrized-type-inheritance
  - https://github.com/mpryahin/kryo-parametrized-type-inheritance/blob/master/src/main/java/io/cleverdata/dmpkit/BaseType.java

**Comment by wangyum** [comment]
retest this please

**Comment by wangyum** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
@wangyum

I made a [PR](https://github.com/wangyum/spark/pull/9) to your branch. Could you review and merge that? The content are the followings.
- Add a new test case for SPARK-23131 (Kryo raises StackOverflow during serializing GLR model)
- Update your test case for SPARK-25176 (ClassCastException when reading/writing a Map)

After merging the [PR](https://github.com/wangyum/spark/pull/9), please add `[SPARK-25176]` after `[SPARK-23131]` because we have test cases. We can resolve both issues under your name.

For the remaining SPARK-20389 (NegativeArraySizeException), there is no response on JIRA. So, let's simply exclude it from this PR description because we cannot claim that it's resolved.

**Comment by wangyum** [comment]
Thanks, @dongjoon-hyun

**Comment by dongjoon-hyun** [comment]
Although this will give us a different Kryo version (not Hive, ORC), the newly added test cases show the benefit clearly. Also, I checked two new test cases with/without this PR. It looks enough to me.
Could you take a look again, @srowen, @cloud-fan and @erikerlandson ?

**Comment by dongjoon-hyun** [comment]
And, @wangyum . Please add `[SPARK-25258]` to the PR title like `[SPARK-25258][SPARK-23131][SPARK-25176]`. SPARK-23131 is the one you created for this PR.

Also, the PR description had better have the following. (We excludes only SPARK-20389.)
```
- SPARK-25258 Upgrade kryo package to version 4.0.2
- SPARK-23131 Kryo raises StackOverflow during serializing GLR model
- SPARK-25176 Kryo fails to serialize a parametrised type hierarchy
```

**Review by srowen — APPROVED** [review]
It's not quite that we shade Kryo, but use a shaded artifact called kryo-shaded. ORC does too, but looks like that can be ruled out as a compatibility issue. I don't see other ways that the version change can leak into user-visible contexts, like through the code or serialized data formats. I can't see evidence of a compatibility issue, as likewise all tests pass.

I think this is OK to merge, given the upside.

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master
